src/handles/handle.rs

```diff
@@ -0,0 +1,330 @@
+    };
+    use log::error;
+    use num_derive::{FromPrimitive, ToPrimitive};
+    use num_traits::{FromPrimitive, ToPrimitive};
+    use std::convert::{From, TryFrom};
+    /// TpmConstant Handle
+    ///
+    /// Represents an esys handle for
+    /// tpm constant resource handles.
+    #[derive(FromPrimitive, ToPrimitive, Debug, Copy, Clone, PartialEq, Eq)]
+    #[repr(u32)]
+    pub enum TpmConstantHandle {
+        Owner = ESYS_TR_RH_OWNER,
+        Null = ESYS_TR_RH_NULL,
+        Lockout = ESYS_TR_RH_LOCKOUT,
+        Endorsement = ESYS_TR_RH_ENDORSEMENT,
+        Platform = ESYS_TR_RH_PLATFORM,
+        PlatformNv = ESYS_TR_RH_PLATFORM_NV,
+    }
+
+    impl_constant_handle_conversion_basic!(TpmConstantHandle);
+    impl_constant_handle_conversion!(TpmConstantHandle, ObjectHandle);
+}
+
+///
+/// Macro for implmeneting conversion between handles.
+///
+macro_rules! impl_handle_conversion {
+    ($handle_type_self:ident, $handle_type_other:ident) => {
+        impl From<$handle_type_self> for $handle_type_other {
+            fn from(handle_self: $handle_type_self) -> $handle_type_other {
+                $handle_type_other::from(ESYS_TR::from(handle_self.value))
+            }
+        }
+
+        impl From<$handle_type_other> for $handle_type_self {
+            fn from(handle_other: $handle_type_other) -> $handle_type_self {
+                $handle_type_self {
+                    value: handle_other.value(),
+                }
+            }
+        }
+    };
+}
+
+/// Auth handle module
+pub mod auth {
+    use super::object::ObjectHandle;
+    impl_basic_handle!(
+        /// Auth handle
+        ///
+        /// Represents an esys handle to resources
+        /// that can be used for authentication.
+        AuthHandle
+    );
+    impl_handle_conversion!(AuthHandle, ObjectHandle);
+}
+
+/// NV Index handle module
+pub mod nv_index {
+    use super::auth::AuthHandle;
+    use super::object::ObjectHandle;
+    impl_basic_handle!(
+        /// NV Index Handle
+        ///
+        /// Represents an esys resource handle
+        /// for a nv index.
+        NvIndexHandle
+    );
+    impl_handle_conversion!(NvIndexHandle, ObjectHandle);
+    impl_handle_conversion!(NvIndexHandle, AuthHandle);
+}
+
+/// Key handle module
+pub mod key {
+    use super::object::ObjectHandle;
+    impl_basic_handle!(
+        /// Key Handle
+        ///
+        /// Represents an esys resource handle
+        /// for a key.
+        KeyHandle
+    );
+    impl_handle_conversion!(KeyHandle, ObjectHandle);
+}
+
+/// Session handle module
+pub mod session {
+    use super::auth::AuthHandle;
+    use super::object::ObjectHandle;
+    impl_basic_handle!(
+        /// Session Handle
+        ///
+        /// Represents an esys handle used for
+        /// referencing session resources.
+        SessionHandle
+    );
+    impl_handle_conversion!(SessionHandle, ObjectHandle);
+    impl_handle_conversion!(SessionHandle, AuthHandle);
```

Comment:
Shouldn't there be a conversion from `PasswordHandle` to `SessionHandle` too?

Reply:
I was thinking of making PasswordHandle and NoneHandles into new types with only one value. In order to be more consistent....